File: leetcode/algorithms/637_average_of_levels_in_binary_tree/main.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var averageOfLevels = function (root) {
  const result = [], queue = [];
  let left = 0, right = 0;

  queue[right++] = root;
  while (left < right) {
    const nodeCount = right - left;
    let sum = 0;
    for (let i = 0; i < nodeCount; i++) {
      const node = queue[left++];
      sum += node.val;
      if (node.left) {
        queue[right++] = node.left;
      }
      if (node.right) {
        queue[right++] = node.right;
      }
    }
    result.push(sum / nodeCount);
  }

  return result;
};

// Solution
var solution = function (root) {
  let q = [root], ans = [];
  while (q.length) {
    let qlen = q.length, row = 0;
    for (let i = 0; i < qlen; i++) {
      let curr = q.shift();
      row += curr.val;
      if (curr.left) {
        q.push(curr.left);
      }
      if (curr.right) {
        q.push(curr.right);
      }
    }
    ans.push(row / qlen);
  }
  return ans;
};

File: leetcode/algorithms/637_average_of_levels_in_binary_tree/main.py
from typing import List
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class AverageOfLevelsInBinaryTree:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        result: List[float] = []
        queue: List[TreeNode] = []

        queue.append(root)
        while len(queue) > 0:
            nodeCount: int = len(queue)
            sum: float = 0
            for i in range(nodeCount):
                node: TreeNode = queue.pop(0)
                sum += node.val
                if node.left is not None:
                    queue.append(node.left)
                if node.right is not None:
                    queue.append(node.right)
            result.append(sum / nodeCount)

        return result

    # Solution
    def solution(self, root: Optional[TreeNode]) -> List[float]:
        q, ans = [root], []
        while len(q):
            qlen, row = len(q), 0
            for i in range(qlen):
                curr = q.pop(0)
                row += curr.val
                if curr.left:
                    q.append(curr.left)
                if curr.right:
                    q.append(curr.right)
            ans.append(row / qlen)
        return ans
